### DISCUSSION TITLE
Caelum ES loads apps with local symbols; breaks CI App

### DISCUSSION THREAD

**Discussion by jcmarsh** [nasa/cFS#388] [Q&A] [answered]
I've been trying out Caelum RC 4, and ran into an issue with apps that load modules: the loaded module can not refer to symbols in the original app.

For example, the [CI app](https://github.com/nasa/CFS_CI) uses modules to implement interfaces such as rs422 and udp. These modules then refer to a symbol in CI. For example, to update a command counter such as [here](https://github.com/nasa/CFS_CI/blob/1bcf88db5880b1983c7cf57ef7bfba42cee5d0e9/fsw/examples/rs422/ci_custom.c#L317)

In previous versions of cFS this was not an issue, as the default behavior in OSAL's OS_ModuleLoad was to load symbols with global visibility. With OSAL 6.0, OS_ModuleLoad now supports two flags, `OS_MODULE_FLAG_GLOBAL_SYMBOLS ` and `OS_MODULE_FLAG_LOCAL_SYMBOLS`, and ES uses local visibility for apps and global visibility for libraries ([relevant code](https://github.com/nasa/cFE/blob/b66661cc6e8f695d5ca9fa56217f58a291d5884d/modules/es/fsw/src/cfe_es_apps.c#L427)).

I am not aware of a way to change this behavior for a specific app.

I don't know the details behind the choice to use local visibility for apps, and the comments for the flags make it sound like there are some very good reasons to do it this way.

Is the use case with CI no longer supported? Should CI be re-written so that the interface modules do not need to link to a symbol in the main CI app? Or is there a way to support this behavior in cFS?

Thank you!

**Comment by skliper** [comment, marked as answer]
CI should be updated, as globally visible symbols are not good for modules that can be unloaded/reloaded/etc.

**Reply by jphickey** [reply]
Yes, just wanted to second this - if symbols need to be accessed from more modules than just itself, it should be a library.

Applications are intended to be reloadable - if you have a dependent module accessing symbols within your application, and the application is unloaded, things will break badly (segfault/crash).  Unfortunately there is no way to know if symbols have been used and/or keep a refcount to make this safe.  The only way to really make it safe is to keep the symbols local/private, then it can be reasonably ensured that nothing else is using any symbols when the app gets unloaded.

If the intent is that the CI app is never unloaded at runtime - then it shouldn't be hard to make the whole thing into a library instead, and just keep the entry point as an app, which calls the main function in the library (or however you choose to divide it).

**Reply by jcmarsh** [reply]
Thank you!

Are there plans for an official update to CI?

**Reply by skliper** [reply]
Ping @tngo67 - do you know if there is an official update planned for CI Caelum interoperability?
